src/main/host/host.rs

```diff
@@ -458,18 +458,53 @@ impl Host {
             trace!("{pid:?} doesn't exist");
             return;
         };
-        let remove_process = {
+        let died;
+        let is_orphan;
+        {
             Worker::set_active_process(&processrc);
             let process = processrc.borrow(self.root());
             process.resume(self, tid);
             Worker::clear_active_process();
-            process.borrow_zombie().is_some() && process.ppid().is_none()
+            died = process.borrow_zombie().is_some();
+            is_orphan = process.parent_id() == ProcessId::INIT;
         };
         RootedRc::explicit_drop(processrc, &self.root);
-        if remove_process {
+
+        if !died {
+            return;
+        }
+
+        // Reparent children, and collect IDs of children that are dead.
+        let mut orphaned_zombie_pids: Vec<ProcessId> = self
+            .processes
+            .borrow()
+            .iter()
+            .filter(|(_pid, processrc)| {
+                let process = processrc.borrow(&self.root);
+                if process.parent_id() != pid {
+                    // Not a child of the current process
+                    false
+                } else {
+                    process.set_parent_id(ProcessId::INIT);
+                    let is_zombie = process.borrow_zombie().is_some();
+                    is_zombie
+                }
+            })
+            .map(|(pid, _process)| *pid)
```

Comment:
nit: `filter_map` instead? (I think could replace `false` with `None` and `is_zombie` with `is_zombie.then_some(pid)`.)